src/utils/leak.js

```diff
@@ -586,7 +586,337 @@ export class LeakImplementation {
+          computeFromValue(buf);
+          return;
+        case "Map":
+        case "WeakMap":
+          ifNotVisited(value, computeFromMap);
+          return;
+        case "Set":
+        case "WeakSet":
+          ifNotVisited(value, computeFromSet);
+          return;
+        default:
+          invariant(kind === "Object", `Object of kind ${kind} is not supported in calls to abstract functions.`);
+          invariant(
+            value.$ParameterMap === undefined,
+            `Arguments object is not supported in calls to abstract functions.`
+          );
+          return;
+      }
+    }
+    function computeFromDescriptor(descriptor: Descriptor): void {
+      invariant(descriptor.value === undefined || descriptor.value instanceof Value);
+      if (descriptor.value !== undefined) computeFromValue(descriptor.value);
+      if (descriptor.get !== undefined) computeFromValue(descriptor.get);
+      if (descriptor.set !== undefined) computeFromValue(descriptor.set);
+    }
+    function computeFromObjectPropertyBinding(binding: PropertyBinding): void {
+      let descriptor = binding.descriptor;
+      if (descriptor === undefined) return; //deleted
+      computeFromDescriptor(descriptor);
+    }
+
+    function computeFromObjectProperties(obj: ObjectValue, kind?: ObjectKind): void {
+      // symbol properties
+      for (let [, propertyBindingValue] of obj.symbols) {
+        invariant(propertyBindingValue);
+        computeFromObjectPropertyBinding(propertyBindingValue);
+      }
+
+      // string properties
+      for (let [, propertyBindingValue] of obj.properties) {
+        invariant(propertyBindingValue);
+        computeFromObjectPropertyBinding(propertyBindingValue);
+      }
+
+      // inject properties with computed names
+      if (obj.unknownProperty !== undefined) {
+        let desc = obj.unknownProperty.descriptor;
+        if (desc !== undefined) {
+          let val = desc.value;
+          invariant(val instanceof AbstractValue);
+          computeFromObjectPropertiesWithComputedNames(val);
+        }
+      }
+
+      // prototype
+      computeFromObjectPrototype(obj);
+    }
+    function computeFromObjectPrototype(obj: ObjectValue) {
+      computeFromValue(obj.$Prototype);
+    }
+    function computeFromFunctionValue(fn: FunctionValue) {
+      computeFromObjectProperties(fn);
+
+      if (fn instanceof BoundFunctionValue) {
+        computeFromValue(fn.$BoundTargetFunction);
+        computeFromValue(fn.$BoundThis);
+        for (let boundArg of fn.$BoundArguments) computeFromValue(boundArg);
+        return;
+      }
+
+      invariant(
+        !(fn instanceof NativeFunctionValue),
+        "all native function values should have already been created outside this pure function"
+      );
+
+      // TODO: Add items to nonLocalReadBindings in passing
+      let nonLocalReadBindings = nonLocalReadBindingsOfFunction(fn);
+      computeFromBindings(fn, nonLocalReadBindings);
```

Comment:
What if there are any `nonLocalReadBindings` left at the end? Shouldn't happen in strict mode, but otherwise there might be some global variable access. What does that mean?

Reply:
I have added an invariant to ensure that it empties out after rewinding down the scopes, and added an issue #2484 to deal with global environments.